third_party/move/move-compiler-v2/src/env_pipeline/inlining_optimization.rs

```diff
@@ -0,0 +1,695 @@
+            // all the callsites of a callee, because we expect to coalesce the locals at
+            // different callsites.
+            // The remaining budget for code size is reduced for each callsite.
+            locals_budget_remaining = locals_reduced;
+            code_size_budget_remaining = code_size_reduced;
+        } else {
+            // cannot inline this callee due to budget limits
+            // maybe try the next candidate
+            continue;
+        }
+    }
+    (
+        call_sites_to_inline,
+        FunctionSize::new(
+            code_size + (code_size_budget - code_size_budget_remaining),
+            num_locals + (locals_budget - locals_budget_remaining),
+        ),
+    )
+}
+
+fn get_function_size_estimate(env: &GlobalEnv, function: &QualifiedId<FunId>) -> FunctionSize {
+    env.function_size_estimate
+        .borrow()
+        .get(function)
+        .copied()
+        .unwrap_or_default()
+}
+
+/// Does `callee` have any privileged operations on structs/enums that cannot be performed
+/// directly in a caller with module id `caller_mid`?
+fn has_privileged_operations(caller_mid: ModuleId, callee: &FunctionEnv) -> bool {
+    let env = callee.env();
+    // keep track if we have found any privileged operations
+    let mut found = false;
+    // used to track if we are within a spec block, privileged operations within
+    // spec blocks are allowed
+    let mut spec_blocks_seen = 0;
+    if let Some(body) = callee.get_def() {
+        body.visit_pre_post(&mut |post, exp: &ExpData| {
+            if !post {
+                if matches!(exp, ExpData::SpecBlock(..)) {
+                    spec_blocks_seen += 1;
+                }
+                if spec_blocks_seen > 0 {
+                    // within a spec block, we can have privileged operations
+                    return true;
+                }
+                // not inside a spec block, see if there are any privileged operations
+                match exp {
+                    ExpData::Call(id, op, _) => match op {
+                        Operation::Exists(_)
+                        | Operation::BorrowGlobal(_)
+                        | Operation::MoveFrom
+                        | Operation::MoveTo => {
+                            let inst = env.get_node_instantiation(*id);
+                            if let Some((struct_env, _)) = inst[0].get_struct(env) {
+                                let struct_mid = struct_env.module_env.get_id();
+                                if struct_mid != caller_mid {
+                                    found = true;
+                                }
+                            }
+                        },
+                        Operation::Select(mid, ..)
+                        | Operation::SelectVariants(mid, ..)
+                        | Operation::TestVariants(mid, ..)
+                        | Operation::Pack(mid, ..) => {
```

Comment:
PackVariant as well

Reply:
There is no `PackVariant` in model ASTs `Operation` (as the final field optionally represents a variant).